Import Pandas and Numpy.

In [1]:
import pandas as pd
import numpy as np

import warnings
warnings.simplefilter('ignore', FutureWarning)

C:\Users\hp\AppData\Local\Temp\ipykernel_18964\1505564052.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


Download the data and read it into a Pandas dataframe.

In [2]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

Check the number of the data points.

In [3]:
concrete_data.shape

(1030, 9)

Check the statistical summary of the data.

In [4]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

Check the data for any missing values.

In [5]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

Split data into target (the column 'Strength') and predictors (the other columns).

In [6]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

In [7]:
predictors.head() # Check the first five rows of the Predictors dataframe

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [8]:
target.head() #Chech the first five values of the Strength column 

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Normalize the data by substracting the mean and dividing by the standard deviation.

In [9]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

Save the number of predictors to *n_cols*.

In [10]:
n_cols = predictors_norm.shape[1] # number of predictors

Import Keras and its required packages.

In [12]:
import keras

In [13]:
from keras.models import Sequential
from keras.layers import Dense

## Build a Neural Network

In [14]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

The above function creates a model with 1 hidden layer and 10 hidden units.


## Train and Test the Network

In [15]:
# build the model
model = regression_model()

C:\Users\hp\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


## Train and and test for 50 cycles (with 100 epochs per cycle) and create a list of 50 mean squared errors representing each cycle.

In [16]:
mean_squared_errors = []
for i in range(50):
    #Randomly split the data (70% for training and 30% for testing), train , and test the model
   
    fitting = model.fit(predictors_norm, target, validation_split=0.3, epochs=100, verbose=0)
    #Find the last mean_squared_error in history.
    mean_squared_error = fitting.history['val_loss'][-1]
    #AAppend every mean_squared_error considered to the list.
    mean_squared_errors.append(mean_squared_error)
    print('i #{}: mean_squared_error {}'.format(i+1, mean_squared_error))

i #1: mean_squared_error 141.12599182128906
i #2: mean_squared_error 90.70285034179688
i #3: mean_squared_error 78.630859375
i #4: mean_squared_error 73.00932312011719
i #5: mean_squared_error 71.47600555419922
i #6: mean_squared_error 78.0284652709961
i #7: mean_squared_error 82.96852111816406
i #8: mean_squared_error 90.44390106201172
i #9: mean_squared_error 101.88128662109375
i #10: mean_squared_error 107.60660552978516
i #11: mean_squared_error 113.03592681884766
i #12: mean_squared_error 116.30558776855469
i #13: mean_squared_error 121.45643615722656
i #14: mean_squared_error 123.49088287353516
i #15: mean_squared_error 125.84979248046875
i #16: mean_squared_error 129.14529418945312
i #17: mean_squared_error 128.57545471191406
i #18: mean_squared_error 129.59451293945312
i #19: mean_squared_error 128.77581787109375
i #20: mean_squared_error 128.79039001464844
i #21: mean_squared_error 129.071533203125
i #22: mean_squared_error 128.8611297607422
i #23: mean_squared_error 132.28630

Report the mean and the standard deviation of the mean squared errors.

In [17]:
print('The mean of the mean squared errors: {}'.format(np.mean(mean_squared_errors)))
print('The standard deviation of the mean squared errors: {}'.format(np.std(mean_squared_errors)))

The mean of the mean squared errors: 131.86137939453124
The standard deviation of the mean squared errors: 25.970897986265886


Investigate the effect of increasing the epochs from 50 to 100.

In [18]:
mean_with_increased_epochs_to_mean= 131.86137939453124/107.7179347229004
mean_with_increased_epochs_to_mean

1.224135792556168

In [19]:
Effect_of_epochs= mean_with_increased_epochs_to_mean-1
Effect_of_epochs

0.224135792556168

Increasing the number of epochs from 50 to 100 has increased the mean of the mean squared errors by 22 %.